### ISSUE TITLE
Produce a warning when map is used instead of flatMap with Callback

### ISSUE THREAD

**Issue by ramnivas** [japgolly/scalajs-react#256]
This is a follow-up from gitter discussion.

Having spent an hour chasing why things didn't work after migration to 0.10.x, I am worried that this class of errors will be a far too easy to make. Consider the following code in a backend class (modified from the actual code to avoid distracting with domain-specific concepts). Here the intention is to update a word list given a new word and its position.

```scala
def updateWord(position: Int, word: String): Callback = 
  scope.props.map { props =>
    val newList = props.wordList.updated(position, word)
    props.onUpdateWordList(newList)
  }
```
`props.onUpdateWordList` is a function defined in its parent component that returns a `Callback`.
I chased why `onUpdateWordList` isn't getting called (after suspecting a lot of other things, first :worried:). Ah, of course, I am using `map`, when I should have used `flatMap`. Far too easy to make such mistakes and no help from scalac.

The core issues is `scalac` converts anything to `Unit`, so what could have been a `Callback[Callback]` was silently treated as `Callback[Unit]` as can be seen from the following reduced example (which compiles fine despite the intention).

```scala
def cb1(): Callback = Callback {}
def cb2(): Callback = Callback {}
def cb1_map_cb2(): Callback = cb1().map{ _ => cb2()}
def cb1_fmap_cb2(): Callback = cb1().flatMap{ _ => cb2()}
```

It will be nice if there can be a warning for `map` in the above example (similar to what is being issued when someone tries to create a `Callback { Callback { } }`.

**Comment by japgolly** [comment]
Thank you very much @ramnivas. I very emphatically want this classes of errors prohibited by scalac. I'll see what I can do.

**Comment by elacin** [comment]
So, I've got a proposal https://github.com/elacin/scalajs-react/commit/3ae014d4c138e4aa46a4cb3872ad8dcf41c41d6e .  

As far as i can tell, the problem is that when `Unit` is expected, scalac happily throws away any value to satisfy that expection:
```scala
scala> val a: Option[Unit] = Some(1)
<console>:10: warning: a pure expression does nothing in statement position; you may be omitting necessary parentheses
       val a: Option[Unit] = Some(1)
                                  ^
a: Option[Unit] = Some(())

scala> 

```
Consequently it's easy to lose a `Callback` (or indeed anything) while mapping. This is not really something we can change.

I propose we change `Callback` from `CallbackTo[Unit]` to `CallbackTo[Empty]`, 
with `Empty` defined something like this:
```scala
sealed trait Empty
val Empty: Empty = new Empty{} //not an object so `Empty.type` cannot be inferred
```

There would be some client breakage, but i would think it's for the better.
In this case we broke `RouterLogic.interpret` - but I dont understand how
it even compiles now.

**Comment by japgolly** [comment]
@elacin Wow that's a really interesting approach! I wasn't thinking along those lines at all. If we went with that approach I think that would solve all the silent-discards-to-`Unit` but it opens up another: 

```scala
val c = CallbackTo(1).map(println) // CallbackTo[Unit]
c: Callback // error saying that CallbackTo[Unit] isn't a CallbackTo[Empty]
```

We can add an implicit conversion from `CallbackTo[Unit]` to `CallbackTo[Empty]` and I *think* (unverified) that will be safe. scalac has a hack where if it knows the desired return type (eg. the return type of a method when the expression is last in the method), it adds towards type inference which enables the throw-everything-away-we-want-`Unit` behaviour. I'm correct that it's safe and there are no other holes, this could be a great solution.

Alternatively, one of the approaches I had in mind was to make methods like `map` a macro that checks the real return type and issues a warning if it's going to be cast to `Unit`.

**Comment by japgolly** [comment]
Also there's a bit of whinging about the existence of `Callback` with the question/complaint being why can't we just use `=> Unit`. I thought it was well enough documented and there's plenty of detail in issues and changelogs and such leading up to the change being made but it doesn't seem to be enough. Now that life has moved on and because I hate repetition, every time I'm asked I provide less of answer. If we break client code making `Callback = CallackTo[Empty]` I think this time we should provide more detail about why/how, probably even provide some examples of the problems that we're solving. Hopefully that will be enough but I'm open to ideas on how we can do better this time.

**Comment by elacin** [comment]
> We can add an implicit conversion from CallbackTo[Unit] to CallbackTo[Empty] and I think (unverified) that will be safe. scalac has a hack where if it knows the desired return type (eg. the return type of a method when the expression is last in the method), it adds towards type inference which enables the throw-everything-away-we-want-Unit behaviour. 

I dont know how scalac works in much detail (though this mirrors my understanding). What i do know is that the inferred type of an expression will vary with the context it's in, so we just need to capture the expression in a context where scalac doesnt have any expectation as to what the type will be.
I think this scheme would have that quality.

> ... but it opens up another ...

This is a major drawback. We would need that implicit conversion, as well as for `CallbackOption[Empty]` and `CallbackOption[Unit]` (one of those already exists). That is complexity that might rear it's head in many unexpected places, I'm sure. 

> Also there's a bit of whinging about the existence of `Callback` ...

There has indeed been some pushback on the whole `Callback` issue (though i would be very interested in hearing what people think after using it for a while! personally I'm a very happy salmon).

That makes it all the more important to not accidentally add even more complexity here.
Case in point - I really don't feel like explaining the difference between `CallbackTo[Unit]` and `CallbackTo[Empty]` to people. If we were to do this it better be pretty transparent. 

> Alternatively, one of the approaches I had in mind was to make methods like map a macro that checks the real return type and issues a warning if it's going to be cast to Unit.

Would like to see this explored as well :+1: .

**Comment by japgolly** [comment]
I don't think we need a macro for this after all. I think we can just use `Callback.ResultGuard` (or a similar trick with implicits) on the `map` method (and any other methods that could have the same potential problem). Too tired to try now though; need sleep. But it should work.

**Comment by japgolly** [comment]
I just had a play with this and this is solvable via a very simple bit of implicit machinery.

```scala
  final class MapGuard[A] { type Out = A }

  object MapGuard {
    private val instance = new MapGuard[Any]
    implicit def id[A]: MapGuard[A] =
      instance.asInstanceOf[MapGuard[A]]
  }

// Then change .map to
  def map[B](g: A => B)(implicit ev: MapGuard[B]): CallbackTo[ev.Out] =
    new CallbackTo(() => g(f()))
```

Then it presents a new question: should we allow any types to be converted to `Callback` automatically? If not, that will force users to call `.void` manually (or add `()` to their callback bodies). I'm thinking maybe it would be ok to allow `CallbackTo[_ <: AnyVal]` to be converted to `Callback`. Thoughts?

**Comment by japgolly** [comment]
Wait, I'm an idiot. We're only talking about `.map`. If one wants to discard the result, they can change `.map(...)` into
*  `.map[Unit](...)`
* or `.map(...: Unit)`

Easy. No uncomfortable implicit conversions necessary. I think we're done!

**Comment by japgolly** [comment]
Closed via 53d649f.
